### ISSUE TITLE
The `–refresh-times` option doesn’t seem to work for an encrypted onedrive remote

### ISSUE THREAD

**Issue by bigtonylewis** [rclone/rclone#8101]
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/the-refresh-times-option-doesnt-seem-to-work-for-an-encrypted-onedrive-remote/47888/1

#### What is the problem you are having with rclone?
The `--refresh-times` option doesn't seem to work for my encrypted onedrive remote.

This is a follow on from my [bug report](https://forum.rclone.org/t/after-upgrade-rclone-sync-wants-to-do-a-full-copy-not-just-of-deltas/47864/8) and subsequent issue #8090.  I reopened the issue but it didn't get visibility

If I sync a single file and then an lsl of the remote file, I can see that it still only has one-second resolution, not millisecond resolution. A subsequent run of sync still reports that it has to adjust the modification time. That is, it appears `--refresh-times` is not working.

#### What is your rclone version (output from `rclone version`)
```
rclone v1.68.0
- os/version: ubuntu 22.04 (64 bit)
- os/kernel: 5.15.0-101-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.23.1
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Ubuntu 22.04.2 LTS

#### Which cloud storage system are you using? (e.g. Google Drive)
Encrypted OneDrive

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
```
rclone -vv sync --refresh-times ~/repos/photos/data/1/1202 onedrive:repos/photos/data/1
```
Then
```
rclone lsl ~/repos/photos/data/1/1202
rclone lsl onedrive:repos/photos/data/1/1202
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
```
borg@host:~$ rclone -vv sync --refresh-times ~/repos/photos/data/1/1202 onedrive:repos/photos/data/1
2024/09/22 23:57:32 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "sync" "--refresh-times" "/home/borg/repos/photos/data/1/1202" "onedrive:repos/pho
tos/data/1"]
2024/09/22 23:57:32 DEBUG : Creating backend with remote "/home/borg/repos/photos/data/1/1202"
2024/09/22 23:57:32 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/22 23:57:32 DEBUG : fs cache: adding new entry for parent of "/home/borg/repos/photos/data/1/1202", "/home/borg/repos/photos/data/1"
2024/09/22 23:57:32 DEBUG : Creating backend with remote "onedrive:repos/photos/data/1"
2024/09/22 23:57:32 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/qtvah5opb18p4g4ve2i0lh4mio/vptothej6ciusdjpca3agsgj5o/u58r1pkbe9ip1vlcarj5tntqqs/s4s1q07jvge4ij8b60qlddrsf8"
2024/09/22 23:57:34 DEBUG : 1202: Modification times differ by -143.88454ms: 2023-02-17 21:00:17.14388454 +1100 AEDT, 2023-02-17 10:00:17 +0000 UTC
2024/09/22 23:57:35 INFO  : 1202: Updated modification time in destination
2024/09/22 23:57:35 DEBUG : 1202: Unchanged skipping
2024/09/22 23:57:35 INFO  :
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Elapsed time:         2.9s

2024/09/22 23:57:35 DEBUG : 6 go routines active

borg@host:~$ rclone lsl ~/repos/photos/data/1/1202
   147857 2023-02-17 21:00:17.143884540 1202
borg@host:~$ rclone lsl onedrive:repos/photos/data/1/1202
   147857 2023-02-17 21:00:17.000000000 1202
```

The key part appears to be that the mtime of the onedrive file still only has precision to a whole second.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
What kind of onedrive are you using?

Can you include the output of `rclone config redacted` please?

I did a test and I'm satisfied that `--refresh-times` is working properly on a onedrive personal wrapped with crypt.

Edited highlights

```
$ rclone sync -vv /tmp/src 'secret,remote="onedrive:secrets":'
$ touch /tmp/src/dir/file.txt
$ rclone sync -vv /tmp/src --refresh-times 'secret,remote="onedrive:secrets":'
INFO  : dir/file.txt: Updated modification time in destination
$ rclone sync -vv /tmp/src 'secret,remote="onedrive:secrets":'
DEBUG : dir/file.txt: Size and modification time the same (differ by -373.028µs, within tolerance 1ms)
DEBUG : dir/file.txt: Unchanged skipping
```

> If I sync a single file and then an lsl of the remote file, I can see that it still only has one-second resolution, not millisecond resolution.

So you uploaded a file but it only has 1 second resolution? And then the resolution did not change when we changed it.

What kind of file is this?

Does it do this for all files? Or just this type of file?

**Comment by bigtonylewis** [comment]
> So you uploaded a file but it only has 1 second resolution? And then the resolution did not change when we changed it.
> What kind of file is this?
> Does it do this for all files? Or just this type of file?
The files are the contents of a borgbackup repository - git-ish incremental backups.  In this context, they are just standard files.  Normally I just do a vanilla `sync`, but this was before doing an upgrade to the latest `rclone`.

Once I upgraded, the large sync was taking too long, and so I investigated.  This mtime mismatch was happening on all the files, forcing a full upload.

To troubleshoot, I devolved the problem into the simplest replicable (for me) use case - that of syncing a single file.  That's what this issue reports against.  For this single file use case, the `--refresh-times` flag doesn't actually seem to refresh the times.

The problem occurs across the whole repository (i.e. doing a sync

**Comment by bigtonylewis** [comment]
rclone config:
```
[onedrive]
type = crypt
remote = onedriveunderlying:rclonebackups
filename_encryption = standard
directory_name_encryption = true
password = XXX
password2 = XXX

[onedriveunderlying]
type = onedrive
token = XXX
drive_id = XXX
drive_type = personal
```

**Comment by bigtonylewis** [comment]
Maybe the difference between your test above and mine is that in mine the files already exist with whole-second precision on the timestamp (created by a pre-1.66 version), whereas yours is going to get the timestamp from the file as it puts it

**Comment by ncw** [comment]
> Maybe the difference between your test above and mine is that in mine the files already exist with whole-second precision on the timestamp (created by a pre-1.66 version), whereas yours is going to get the timestamp from the file as it puts it

That worked for me - here the upload was made with v1.65.2 and you can see v1.68.1 correcting the times with `--refresh-times`

```
$ rclone-v1.65.2 sync -vv /tmp/src 'secret,remote="onedrive:secrets":'
$ rclone lsl 'secret,remote="onedrive:secrets":'
        6 2024-09-25 15:14:39.000000000 dir/file.txt
  3538536 2024-09-25 12:23:49.000000000 dir/test.jpg
$ rclone-v1.68.1 sync -vv /tmp/src --refresh-times 'secret,remote="onedrive:secrets":'
INFO  : dir/file.txt: Updated modification time in destination
INFO  : dir/test.jpg: Updated modification time in destination
$ rclone lsl 'secret,remote="onedrive:secrets":'
        6 2024-09-25 15:14:39.268000000 dir/file.txt
  3538536 2024-09-25 12:23:49.422000000 dir/test.jpg
```

I wonder whether these old files are different somehow and can't take a ms precision timestamp, but only a 1s timestamp precision.

You could see if new files can take ms precision. Can you try uploading a new file and try the `--refresh-times` on a sync after `touch`-ing the local one?

**Comment by sweharris** [comment]
OK, on _one_ of my remotes this is happening again and again, so I'm able to get a debug log.

```
2024/09/25 16:33:57 DEBUG : rclone: Version "v1.68.0" starting with parameters ["/home/sweh/bin/rclone" "--config" "/home/sweh/.config/rclone/rclone.conf" "--delete-before" "--rc" "--log-level" "DEBUG" "--log-file" "/BACKUP/NO_SYNC/log/20240925_OneDriveBackups3:.txt" "--stats=60s" "sync" "." "OneDriveBackups3:" "--include" "brains/**" "--include" "monitor/**" "--include" "openwrt/**" "--include" "pugwash/**" "--include" "switches/**" "--include" "tplink-a7/**" "--include" "monitor9/**" "--include" "hass9/**" "--include" "plexserver9/**" "--include" "phoenix/**"]
```

It starts off good...

```
2024/09/25 16:34:02 DEBUG : switches/20200720-basement-switch.cfg: Size and modification time the same (differ by 0s, within tolerance 1ms)
2024/09/25 16:34:02 DEBUG : switches/20200720-basement-switch.cfg: Unchanged skipping
2024/09/25 16:34:02 DEBUG : switches/20200720-living-switch.cfg: Size and modification time the same (differ by 0s, within tolerance 1ms)
```

But then a little later...

```
2024/09/25 16:34:03 DEBUG : brains/20240829._boot.4.gz_aa: Modification times differ by -909.387934ms: 2024-08-29 00:15:45.909387934 -0400 EDT, 2024-08-29 04:15:45 +0000 UTC
2024/09/25 16:34:03 DEBUG : brains/20240829.log: Modification times differ by -68.388101ms: 2024-08-29 00:15:46.068388101 -0400 EDT, 2024-08-29 04:15:46 +0000 UTC
2024/09/25 16:34:03 DEBUG : brains/20240829.root.4.gz_aa: Modification times differ by -257.383046ms: 2024-08-29 00:15:41.257383046 -0400 EDT, 2024-08-29 04:15:41 +0000 UTC
```

Let's look at the last of these:

```
% stat /BACKUP/brains/20240829.root.4.gz_aa                      
  File: /BACKUP/brains/20240829.root.4.gz_aa
  Size: 140938358       Blocks: 275552     IO Block: 4096   regular file
Device: fd05h/64773d    Inode: 81166355    Links: 1
Access: (0660/-rw-rw----)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-09-25 14:19:52.965512601 -0400
Modify: 2024-08-29 00:15:41.257383046 -0400
Change: 2024-08-29 00:15:42.569384424 -0400
 Birth: 2024-08-29 00:15:06.024346027 -0400
```

This is looking like OneDrive is only doing 1 second accuracy, and _not_ millisecond.

(Why didn't that show on the switch files?  Huh stat shows .0000 as the last modified time; I guess that must have originally been written on a filesystem without ms accuracy and just copied over!)

What's odd is that the other backups (OneDriveBackup1:  and OneDriveBackup2:) don't show this!

The 3 crypt's are set up the same
```
OneDriveBackups]
type = crypt
remote = OneDriveSweharris2:Backups
filename_encryption = off
directory_name_encryption = false
password = XXX
password2 = XXX

[OneDriveBackups2]
type = crypt
remote = OneDriveSweharris3:Backups
filename_encryption = off
directory_name_encryption = false
password = XXX
password2 = XXX

[OneDriveBackups3]
type = crypt
remote = OneDriveSweharris:Onedrive Only/BACKUP3
filename_encryption = off
directory_name_encryption = false
password = XXX
password2 = XXX
```

The underlying remotes are similar:

```
[OneDriveSweharris]
type = onedrive
token = ...
drive_id = ...
drive_type = personal

[OneDriveSweharris2]
type = onedrive
token = ...
drive_id = ...
drive_type = personal

[OneDriveSweharris3]
type = onedrive
token = ...
drive_id = ...
drive_type = personal
```

The only difference _might_ be in the creation of the OneDrive itself.  OneDriveSweharris1 was my original "created with outlook" drive.  I then converted that to a family plan.  "2" and "3" were created with the family plan.  This _shouldn't_ make a difference, but it's the only one I can see!

Now looking at a new file uploaded today using `rclone lsl ...`

```
 61238918 2024-09-25 00:15:21.000000000 20240925.root.3.gz_aa
```

So this brand new file still only has one second accuracy!

**Comment by ncw** [comment]
Hmm, so some OneDrive personal accounts apparently can't take ms accuracy timestamps apparently.

I wonder if we can tell them apart somehow.

@sweharris can you try some listings with -vv --dump bodies and see if you can spot any differences between the 1s OneDrive and the 1ms one?

@nielash you put the ms timestamps in - I wonder if you have any ideas about what is going on.

I guess we could set the precision back to 1s - that wouldn't be a big loss.

**Comment by sweharris** [comment]
There's a lot of differences..

For brains (the "broken" one) I see something massively long graph downloadURL

```
      "@microsoft.graph.downloadUrl": "https://my.microsoftpersonalcontent.com/p
ersonal/XXX/_layouts/15/download.aspx?UniqueId=XXX&Translate=false&tempauth=XXX&ApiVersion=2.0",
```

For "penfold" (good)
```
      "@microsoft.graph.downloadUrl": "https://straqa.DM.files.1drv.com/XXX",
```

The corrending webURLs are also different

brains: `      "webUrl": "https://onedrive.live.com?cid=XXX",`

penfold: `      "webUrl": "https://1drv.ms/u/XXX",`

Creator is also different; for "brains" it's "Microsoft Graph"
```
      "createdBy": {
        "application": {
          "id": "00000003-0000-0000-c000-000000000000",
          "displayName": "Microsoft Graph"
        },
```

 but for "penfold" it's "rclone"

```
      "createdBy": {
        "application": {
          "displayName": "rclone",
          "id": "48211038"
        },
```

Parent reference is also different:

brains:
```
      "parentReference": {
        "driveType": "personal",
        "driveId": "XXX",
        "id": "XXX",
        "name": "brains",
        "path": "/drives/XXX/root:/Onedrive Only/BACKUP3/brains",
        "siteId": "XXX"
      },
```

penfold:
```
      "parentReference": {
        "driveId": "XXX",
        "driveType": "personal",
        "id": "XXX",
        "name": "penfold",
        "path": "/drive/root:/Backups/penfold"
      },
```

(Interestingly, but not relevent to this problem, brains has quickXorHash, penfold has quickXorHash _and_ sha1Hash _and_ sha256Hash).

These look like two totally different backends!!

**Comment by bigtonylewis** [comment]
Still the same problem.

Sync the file to a new subdir on the remote
```
borg@host:~$ rclone -vv sync ~/repos/photos/data/1/1202 onedrive:test/
2024/09/26 08:06:54 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "sync" "/home/borg/repos/photos/data/1/1202" "onedrive:test/"]
2024/09/26 08:06:54 DEBUG : Creating backend with remote "/home/borg/repos/photos/data/1/1202"
2024/09/26 08:06:54 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:06:54 DEBUG : fs cache: adding new entry for parent of "/home/borg/repos/photos/data/1/1202", "/home/borg/repos/photos/data/1"
2024/09/26 08:06:54 DEBUG : Creating backend with remote "onedrive:test/"
2024/09/26 08:06:54 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
2024/09/26 08:06:57 DEBUG : 1202: Need to transfer - File not found at Destination
2024/09/26 08:06:58 DEBUG : vknf9rh5le44k0v06gdv2ri0ts: Starting multipart upload
2024/09/26 08:06:59 DEBUG : vknf9rh5le44k0v06gdv2ri0ts: Uploading segment 0/147937 size 147937
2024/09/26 08:07:00 DEBUG : 1202: quickxor = 8984d4cb626e8166c6c4a481ccedcb1927e8b494 OK
2024/09/26 08:07:00 INFO  : 1202: Copied (new)
2024/09/26 08:07:00 INFO  : 
Transferred:      144.470 KiB / 144.470 KiB, 100%, 48.156 KiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:         6.5s

2024/09/26 08:07:00 DEBUG : 10 go routines active
```
Check the timestamp; it has whole-second resolution:
```
borg@host:~$ rclone -vv lsl onedrive:test/1202
2024/09/26 08:07:12 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "lsl" "onedrive:test/1202"]
2024/09/26 08:07:12 DEBUG : Creating backend with remote "onedrive:test/1202"
2024/09/26 08:07:12 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:07:12 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c/vknf9rh5le44k0v06gdv2ri0ts"
2024/09/26 08:07:15 DEBUG : fs cache: adding new entry for parent of "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c/vknf9rh5le44k0v06gdv2ri0ts", "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
2024/09/26 08:07:15 DEBUG : fs cache: adding new entry for parent of "onedrive:test/1202", "onedrive:test"
   147857 2023-02-17 21:00:17.000000000 1202
2024/09/26 08:07:15 DEBUG : 6 go routines active
```
And sync again anyway.  It still uploads the whole file:
```
borg@host:~$ rclone -vv sync ~/repos/photos/data/1/1202 onedrive:test/
2024/09/26 08:07:22 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "sync" "/home/borg/repos/photos/data/1/1202" "onedrive:test/"]
2024/09/26 08:07:22 DEBUG : Creating backend with remote "/home/borg/repos/photos/data/1/1202"
2024/09/26 08:07:22 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:07:22 DEBUG : fs cache: adding new entry for parent of "/home/borg/repos/photos/data/1/1202", "/home/borg/repos/photos/data/1"
2024/09/26 08:07:22 DEBUG : Creating backend with remote "onedrive:test/"
2024/09/26 08:07:22 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
2024/09/26 08:07:24 DEBUG : 1202: Modification times differ by -143.88454ms: 2023-02-17 21:00:17.14388454 +1100 AEDT, 2023-02-17 10:00:17 +0000 UTC
2024/09/26 08:07:24 DEBUG : vknf9rh5le44k0v06gdv2ri0ts: Starting multipart upload
2024/09/26 08:07:25 DEBUG : vknf9rh5le44k0v06gdv2ri0ts: Uploading segment 0/147937 size 147937
2024/09/26 08:07:26 DEBUG : 1202: quickxor = 6848283a5d1f1294b70f89e75bd1d5b447077fc7 OK
2024/09/26 08:07:26 INFO  : 1202: Copied (replaced existing)
2024/09/26 08:07:26 INFO  : 
Transferred:      144.470 KiB / 144.470 KiB, 100%, 144.473 KiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:         3.6s

2024/09/26 08:07:26 DEBUG : 9 go routines active
```
On the off chance something was caching the file, I did it with a brand new file, and still got the same whole-second resolution:
```
borg@host:~$ date > testfile
borg@host:~$ rclone -vv sync ~/testfile onedrive:test/
2024/09/26 08:07:53 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "sync" "/home/borg/testfile" "onedrive:test/"]
2024/09/26 08:07:53 DEBUG : Creating backend with remote "/home/borg/testfile"
2024/09/26 08:07:53 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:07:53 DEBUG : fs cache: adding new entry for parent of "/home/borg/testfile", "/home/borg"
2024/09/26 08:07:53 DEBUG : Creating backend with remote "onedrive:test/"
2024/09/26 08:07:53 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
2024/09/26 08:07:54 DEBUG : testfile: Need to transfer - File not found at Destination
2024/09/26 08:07:54 DEBUG : jgnrvhn2i0lfrs5req4m822014: Starting multipart upload
2024/09/26 08:07:55 DEBUG : jgnrvhn2i0lfrs5req4m822014: Uploading segment 0/81 size 81
2024/09/26 08:07:56 DEBUG : testfile: quickxor = aed928320bb991ef7a5db7e8d2e807ec70366e51 OK
2024/09/26 08:07:56 INFO  : testfile: Copied (new)
2024/09/26 08:07:56 INFO  : 
Transferred:             81 B / 81 B, 100%, 81 B/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:         3.5s

2024/09/26 08:07:56 DEBUG : 9 go routines active
borg@host:~$ rclone -vv lsl onedrive:test/
2024/09/26 08:08:03 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "lsl" "onedrive:test/"]
2024/09/26 08:08:03 DEBUG : Creating backend with remote "onedrive:test/"
2024/09/26 08:08:03 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:08:03 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
   147857 2023-02-17 21:00:17.000000000 1202
       33 2024-09-26 08:07:44.000000000 testfile
2024/09/26 08:08:05 DEBUG : 6 go routines active
borg@host:~$ rclone -vv sync ~/testfile onedrive:test/
2024/09/26 08:08:12 DEBUG : rclone: Version "v1.68.0" starting with parameters ["rclone" "-vv" "sync" "/home/borg/testfile" "onedrive:test/"]
2024/09/26 08:08:12 DEBUG : Creating backend with remote "/home/borg/testfile"
2024/09/26 08:08:12 DEBUG : Using config file from "/home/borg/.config/rclone/rclone.conf"
2024/09/26 08:08:12 DEBUG : fs cache: adding new entry for parent of "/home/borg/testfile", "/home/borg"
2024/09/26 08:08:12 DEBUG : Creating backend with remote "onedrive:test/"
2024/09/26 08:08:12 DEBUG : Creating backend with remote "onedriveunderlying:rclonebackups/fumo5lkns3qomebaj3e1jacn2c"
2024/09/26 08:08:13 DEBUG : testfile: Modification times differ by -429.469948ms: 2024-09-26 08:07:44.429469948 +1000 AEST, 2024-09-25 22:07:44 +0000 UTC
2024/09/26 08:08:13 DEBUG : jgnrvhn2i0lfrs5req4m822014: Starting multipart upload
2024/09/26 08:08:14 DEBUG : jgnrvhn2i0lfrs5req4m822014: Uploading segment 0/81 size 81
2024/09/26 08:08:15 DEBUG : testfile: quickxor = 472be3848eda28a048aa5a57107a57782ffe8cba OK
2024/09/26 08:08:15 INFO  : testfile: Copied (replaced existing)
2024/09/26 08:08:15 INFO  : 
Transferred:             81 B / 81 B, 100%, 80 B/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:         2.8s

2024/09/26 08:08:15 DEBUG : 9 go routines active
borg@host:~$ 
```

**Comment by sweharris** [comment]
Looking at the original `GET /v1.0/drives/XXX/root` call I also see some differences.

penfold (1ms) has a cTag in the JSON, brains doesn't.

webURL for brains (1s) also has an `id` field:

brains: `"webUrl": "https://onedrive.live.com?cid=XXX&id=YYY",
penfold: `"webUrl": "https://onedrive.live.com/?cid=XXX",`

penfold has additional data in the JSON

```
  "createdBy": {
    "application": {
      "displayName": "OneDrive website",
      "id": "44048800"
    },
```

eTag on brains: `  "eTag": "\"{GUID},139\"",`
on penfold: `  "eTag": "XXX",`

In all cases (both the "root" and the file above), all timestamps for penfold have SS.uuuZ microseconds, but brains is just SSZ second accuracy.
eg
`  "lastModifiedDateTime": "2024-09-25T20:33:55.187Z",`
vs
`  "lastModifiedDateTime": "2024-09-25T21:15:11Z",`

**Comment by bigtonylewis** [comment]
Interestingly, similar to @sweharris I have a second encrypted OneDrive remote, and this problem does not occur on it.

If I run with `--dump bodies` I see that the (1s) repo has a very recent creation time:
```
  "fileSystemInfo": {
    "createdDateTime": "2024-08-27T14:24:00Z",
    "lastModifiedDateTime": "2024-09-25T22:08:15Z"
  },
```
But it's years old to me.  The same element from the (1ms) drive shows a timestamp of 2014.

I wonder if they've migrated this drive to some new platform behind the scenes.  And if so, this problem might be coming for others too, as they migrate more drives over.

**Comment by sweharris** [comment]
Interesting, my 1s also has a recent createdDateTime: `2024-09-22T13:48:05Z` whereas the 1ms one looks more accurate: `2020-01-23T18:42:01.63Z`.

Could well be backend migrations occuring!

That would also explain why I started seeing this problem again a couple of days ago!  Something changed.

**Comment by nielash** [comment]
I am stumped. The migration theory above sounds as plausible as anything. While working on this I found many other undocumented oddities in the API.

**Comment by ncw** [comment]
> (Interestingly, but not relevent to this problem, brains has quickXorHash, penfold has quickXorHash _and_ sha1Hash _and_ sha256Hash).
> 
> These look like two totally different backends!!

That is interesting info. penfold (1ms) having those hashes means it is based off the onedrive personal line. Whereas brains (1s) looks like it is based off onedrive business which never had sha1 or sha256. However Microsoft have deprecated the sha1 and sha256 hashes so migrating to a platform that doesn't have them makes sense. My guess is that they are trying to merge the OneDrive Personal and Business infrastructure.

I've noticed before that Microsoft's infrastructure (not surprisingly for something so big) is not homogeneous.

So if Microsoft are migrating OneDrive accounts to backends which only support 1s time resolution then we should probably revert the 1ms change.

I don't think that 1ms time accuracy is documented anywhere.

@nielash did you find documentation for it somewhere? The API docs I've see are all a bit vague about what a timestamp is.

If we do revert the 1 ms precision can anyone think of any negative consequences. What about bisync?

**Comment by sweharris** [comment]
For testing purposes I added `--modify-window 1s` to my calls, and it worked as expected for my `sync` calls.  I don't use `bisync` but I can't think it'd break anything; anything that was more than 1ms different would already trigger a copy and I can't imagine someone changing a file 10 times a second and expecting each copy to get sync'd; it takes longer than that to copy the file!

**Comment by nielash** [comment]
> @nielash did you find documentation for it somewhere? The API docs I've see are all a bit vague about what a timestamp is.

I agree that the docs are vague, and I would also call them inconsistent. For instance, [this example](https://learn.microsoft.com/en-us/graph/api/driveitem-list-versions?view=graph-rest-1.0&tabs=http#response-1) shows a `lastModifiedDateTime` with 3 decimal places (`2017-09-14T12:34:53.912Z`), while [this one](https://learn.microsoft.com/en-us/graph/api/driveitem-get?view=graph-rest-1.0&tabs=http#response-1) shows no decimal.

> If we do revert the 1 ms precision can anyone think of any negative consequences. What about bisync?

Bisync should be fine -- in theory it should be the same as just manually overriding the precision with `--modify-window 1s`. There's a theoretical risk that two files could be considered equal when they are not, but in practice I think the risk would be very small, as the files would have to have an identical size and modtimes that differ by a fraction of a second.

One thing I wonder is what happens to a file that's already on onedrive with a more precise modtime before this migration. Does the existing decimal data disappear in the API response? Or does it only affect files created/updated after the migration? The former could possibly cause surprises if people are (for example) scripting with the output of `lsl`, or something.

**Comment by sweharris** [comment]
FWIW, on my "bad" drive, an `rclone lsl` of the OneDrive (not even the crypt overlay but the underlying onedrive) shows `.000000000` on every file.  This definitely had files on it before the migration.

Looking at the `dump bodies` output, all the API results appear to be limited to seconds only (eg `"lastModifiedDateTime": "2022-04-07T20:37:14Z"`)

Looking at the complete `lsl` dump output there are only _two_ entries that have a `.` in them, and that's in the `takenDateTime` field.  And that's for 2 new images I uploaded from the Android OneDrive app today; and even there it's only centiseconds; `"takenDateTime": "2024-08-29T18:58:45.56Z"`

(Possibly humorous relation; the Android OneDrive app stopped uploading images and I had to clear data before it would start again; makes me wonder if this backend migration also broke Microsoft's own software!)

**Comment by ncw** [comment]
Thanks for all your testing and investigations.

I've decided to revert the precision back to 1s - I don't think there are any downsides and this should fix the problem - please give this a quick test.

[v1.69.0-beta.8342.4497e8c9c.fix-8101-onedrive-precision](https://beta.rclone.org/branch/fix-8101-onedrive-precision/v1.69.0-beta.8342.4497e8c9c.fix-8101-onedrive-precision/) on branch [fix-8101-onedrive-precision](https://github.com/rclone/rclone/tree/fix-8101-onedrive-precision) (uploaded in 15-30 mins)

**Comment by bigtonylewis** [comment]
I'll be unable to check this for a couple of weeks, but thank you

**Comment by sweharris** [comment]
It _seems_ to be OK.  I removed the `--modify-window 1s` I had before, so now it's using native precision.

eg
```
/home/sweh/bin/rclone.tst --config /home/sweh/.config/rclone/rclone.conf --delete-before  -v --stats=60s sync . OneDriveBackups3:
```

The complete run:

```
% sudo ./SYNC.onedrive.tst

Fri Sep 27 10:16:40 EDT 2024 Starting OneDriveBackups:
2024/09/27 10:16:43 INFO  : SYNC.onedrive.tst: Copied (replaced existing)
2024/09/27 10:16:44 INFO  :
Transferred:        1.847 KiB / 1.847 KiB, 100%, 1.846 KiB/s, ETA 0s
Checks:               471 / 471, 100%
Transferred:            1 / 1, 100%
Elapsed time:         3.8s

Fri Sep 27 10:16:44 EDT 2024 Ended OneDriveBackups:

Fri Sep 27 10:16:44 EDT 2024 Starting OneDriveBackups2:
2024/09/27 10:16:48 INFO  : There was nothing to transfer
2024/09/27 10:16:48 INFO  :
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:               742 / 742, 100%
Elapsed time:         3.7s

Fri Sep 27 10:16:48 EDT 2024 Ended OneDriveBackups2:

Fri Sep 27 10:16:48 EDT 2024 Starting OneDriveBackups3:
2024/09/27 10:16:52 INFO  : There was nothing to transfer
2024/09/27 10:16:52 INFO  :
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:               459 / 459, 100%
Elapsed time:         4.3s

Fri Sep 27 10:16:52 EDT 2024 Ended OneDriveBackups3:

Cloud space:   
OneDriveBackups:        OneDriveBackups2:       OneDriveBackups3:
====================    ====================    ====================
Total:   1.005 TiB      Total:   1.005 TiB      Total:   1.005 TiB
Used:    994.400 GiB    Used:    770 GiB        Used:    249.100 GiB
Free:    34.600 GiB     Free:    259 GiB        Free:    779.900 GiB
Trashed: 1.200 TiB      Trashed: 198.905 GiB    Trashed: 0 B
```

So it successfully sync'd on both types of OneDrive backend.
This is what I expected because the morning sync had already occured, so there shouldn't be any files to transfer (except this one test file).

**Comment by ncw** [comment]
Thanks for testing :-)

**Comment by kapitainsky** [comment]
I have had the same issue and adding `--modify-window 1s` solved it - and I forgot about it. The problem was crazy as not all files were re-synced in my rclone union. I had no time to investigate especially that solution existed.

When reading this thread I looked closer at different OneDrive accounts I have and noticed that at least their GUI differs. For example:

<img width="890" alt="Screenshot 2024-09-27 at 8 16 52 pm" src="https://github.com/user-attachments/assets/e8012db7-5e3b-4b16-8c26-533ad5c282e6">

<img width="890" alt="Screenshot 2024-09-27 at 8 17 26 pm" src="https://github.com/user-attachments/assets/fa58ce9c-b157-4b9b-bd65-586fb850a53d">

These two accounts were created on the same day few months ago - still it looks like they use different OneDrive software:) If GUI differs then API can differ too:)

**Comment by ncw** [comment]
Thanks for your observations @kapitainsky  and your testing @sweharris 

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.69 and I'll put it in 1.68.2 if we make one of those.

**Comment by MistarMuffin** [comment]
@ncw Did this get left out of 1.68.2?
